controlplane/kubeadm/internal/machine_filters.go

```diff
@@ -93,10 +94,34 @@ func OwnedControlPlaneMachines(controlPlaneName string) MachineFilter {
 		if controllerRef == nil {
 			return false
 		}
-		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == controlPlaneName
+		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == owner.GetName() && controllerRef.UID == owner.GetUID()
```

Comment:
That is a very good point. I don't think the adoption code _will_ reconcile the UID properly as it's currently written, but maybe `clusterctl move` knows how to handle that already? Otherwise, I'd expect the GC to cause trouble for pivoting anyway.

Reply:
It appears that clusterctl move does update owner references as objects are moved across: https://github.com/kubernetes-sigs/cluster-api/blob/5154da4caf18b291b1e1e1aba0c635e1d76d3093/cmd/clusterctl/client/cluster/mover.go#L458

That is an impressive piece of code!

I think what that means for us is that UIDs are valid in post-clusterctl v2 workflows. Is there a support matrix where we might need to consider a clusterctl v1 pivot with a KCP?

Reply:
I don't think we necessarily need to consider the v1 pivot, but I do think that we need to consider the use case of someone backing up and restoring resources from backup, which I'm not sure that tools in that space would necessarily do the right thing with the UID there.

It seems like the controller-runtime code handles overwriting the controller ref if the api version or the UID are out of date, and I suspect we should likely do the same (even if part of a different workflow than the comparison here)